### ISSUE TITLE
bug(kartograf): ONNX export parity check uses dot product, comment claims cosine

### ISSUE THREAD

**Issue by Memphis-Chains** [Memphis-Chains/memphis#273]
## Severity
P2 — correctness, future-proofing

## Problem
\`tools/training/kartograf_train/onnx_export.py:178-179\`:

\`\`\`python
# Cosine similarity on embeddings — should be ~1.0 since both are
# L2-normalized and computed on the same weights.
torch_emb_np = torch_emb.detach().cpu().numpy().astype(np.float64)
ort_emb_np = ort_emb.astype(np.float64)
cos = float(np.sum(torch_emb_np * ort_emb_np, axis=-1).mean())
\`\`\`

The variable \`cos\` and the comment claim "cosine similarity," but the math computes a **raw dot product** (\`sum(a*b)\` without dividing by \`||a|| · ||b||\`). For L2-normalized vectors these are mathematically equivalent — and Kartograf's export wrapper does L2-normalize the embedding head's output (\`onnx_export.py:77-78\`). So **today** the parity check is correct.

**Why this is still a bug worth filing:** if anyone in the future removes the L2-normalize from the export wrapper (e.g., Phase 4b INT8 quantization variant where normalization is folded differently, or a Kartograf v2 architecture experiment), the parity check will silently keep passing even when torch and ONNX outputs diverge in magnitude. A bad checkpoint could ship.

## Proposed fix
Two options:

**A. Defensive: compute true cosine explicitly.**
\`\`\`python
def _cosine(a: np.ndarray, b: np.ndarray) -> float:
    a_norms = np.linalg.norm(a, axis=-1, keepdims=True).clip(min=1e-12)
    b_norms = np.linalg.norm(b, axis=-1, keepdims=True).clip(min=1e-12)
    return float(((a / a_norms) * (b / b_norms)).sum(axis=-1).mean())
cos = _cosine(torch_emb_np, ort_emb_np)
\`\`\`

**B. Minimal: rename + assert L2-norm precondition.** Keep the dot-product math (it's cheaper) but rename \`cos\` → \`dot_product\` and add an assertion that both inputs are L2-normalized at the call site. Still future-safe because the assertion fires when normalization is removed.

Recommend **A** — explicit cosine is more robust against subtle future changes (e.g., partial normalization, different epsilon clamping in ONNX vs torch).

## Verification
- After fix: existing tests still pass (current L2-normalized embeddings).
- Negative test: forge non-normalized embedding tensors, assert the cosine check still gives correct value (\~1.0 when same direction even if different magnitudes; bad parity gets caught).

## Source
Codex review forwarded from PR #269 by session 2 — disjoint scope, handed back to session 1 (kartograf trainer). Will be addressed in PR #268 follow-up commit after current training run completes.
